ome_zarr/data.py

```diff
@@ -0,0 +1,166 @@
+from skimage.segmentation import clear_border
+
+from .conversions import rgba_to_int
+from .scale import Scaler
+
+CHANNEL_DIMENSION = 1
+
+
+def coins() -> Tuple[List, List]:
+    """Sample data from skimage."""
+    # Thanks to Juan
+    # https://gist.github.com/jni/62e07ddd135dbb107278bc04c0f9a8e7
+    image = data.coins()[50:-50, 50:-50]
+    thresh = threshold_otsu(image)
+    bw = closing(image > thresh, square(4))
+    cleared = remove_small_objects(clear_border(bw), 20)
+    label_image = label(cleared)
+
+    pyramid = list(reversed([zoom(image, 2 ** i, order=3) for i in range(4)]))
+    labels = list(reversed([zoom(label_image, 2 ** i, order=0) for i in range(4)]))
+
+    pyramid = [rgb_to_5d(layer) for layer in pyramid]
+    labels = [rgb_to_5d(layer) for layer in labels]
+    return pyramid, labels
+
+
+def astronaut() -> Tuple[List, List]:
+    """Sample data from skimage."""
+    scaler = Scaler()
+
+    pixels = rgb_to_5d(np.tile(data.astronaut(), (2, 2, 1)))
+    pyramid = scaler.nearest(pixels)
+
+    shape = list(pyramid[0].shape)
+    shape[1] = 1
+    label = np.zeros(shape)
+    make_circle(100, 100, 1, label[0, 0, 0, 200:300, 200:300])
+    make_circle(150, 150, 2, label[0, 0, 0, 250:400, 250:400])
+    labels = scaler.nearest(label)
+
+    return pyramid, labels
+
+
+def make_circle(h: int, w: int, value: int, target: np.ndarray) -> None:
+    """Apply a 2D circular mask to the given array.
+
+    >>> import numpy as np
+    >>> example = np.zeros((8, 8))
+    >>> make_circle(8, 8, 1, example)
+    >>> print(example)
+    [[0. 0. 0. 0. 0. 0. 0. 0.]
+     [0. 0. 1. 1. 1. 1. 1. 0.]
+     [0. 1. 1. 1. 1. 1. 1. 1.]
+     [0. 1. 1. 1. 1. 1. 1. 1.]
+     [0. 1. 1. 1. 1. 1. 1. 1.]
+     [0. 1. 1. 1. 1. 1. 1. 1.]
+     [0. 1. 1. 1. 1. 1. 1. 1.]
+     [0. 0. 1. 1. 1. 1. 1. 0.]]
+    """
+    x = np.arange(0, w)
+    y = np.arange(0, h)
+
+    cx = w // 2
+    cy = h // 2
+    r = min(w, h) // 2
+
+    mask = (x[np.newaxis, :] - cx) ** 2 + (y[:, np.newaxis] - cy) ** 2 < r ** 2
+    target[mask] = value
+
+
+def rgb_to_5d(pixels: np.ndarray) -> List:
+    """Convert an RGB image into 5D image (t, c, z, y, x)."""
+    if len(pixels.shape) == 2:
+        stack = np.array([pixels])
+        channels = np.array([stack])
+    elif len(pixels.shape) == 3:
+        size_c = pixels.shape[2]
+        channels = [np.array([pixels[:, :, c]]) for c in range(size_c)]
+    else:
+        assert False, f"expecting 2 or 3d: ({pixels.shape})"
+    video = np.array([channels])
+    return video
+
+
+def write_multiscale(pyramid: List, group: zarr.Group) -> None:
+    """Write a pyramid with multiscale metadata to disk."""
+    paths = []
+    for path, dataset in enumerate(pyramid):
+        group.create_dataset(str(path), data=pyramid[path])
+        paths.append({"path": str(path)})
+
+    multiscales = [{"version": "0.1", "datasets": paths}]
+    group.attrs["multiscales"] = multiscales
```

Comment:
This made me wonder where this logic should live in future improvements of this API. The two immediate options I can think of would be:

- have each `Spec` subclass implementing its own serialization method `Multiscales.write(paths)`
- have an `ome_zarr.writer` module mirroring the reader one
- have a

Reply:
Yeah, there is definitely more refactoring and encapsulating to be done: I just had to draw the line somewhere. Hopefully at this point it's fairly straightforward to see where something should live (in this case I'd say the `Spec` subclass and if we want to then wrap that with a Writer we can).

Reply:
:+1: not a blocker. My initial feeling would also be the `Spec` subclass which might lead us to moving it to a separate module e.g. `ome_zarr.specifications` that could be consumed both by readers and writers